server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -3469,8 +3469,31 @@ private void doSnapshotShard(SnapshotShardContext context) {
     private static void ensureNotAborted(ShardId shardId, SnapshotId snapshotId, IndexShardSnapshotStatus snapshotStatus, String fileName) {
         try {
             snapshotStatus.ensureNotAborted();
+
+            if (snapshotStatus.getStage() != IndexShardSnapshotStatus.Stage.INIT
+                && snapshotStatus.getStage() != IndexShardSnapshotStatus.Stage.STARTED) {
+                // A normally running shard snapshot should be in stage INIT or STARTED. And we know it's not in PAUSING or ABORTED because
+                // the ensureNotAborted() call above did not throw. The remaining options don't make sense, if they ever happen.
+                logger.info(
+                    "Shard snapshot found no interrupt signal, but is in an unexpected state. ShardId [{}], SnapshotID [{}], Stage [{}]",
+                    shardId,
+                    snapshotId,
+                    snapshotStatus.getStage()
```

Comment:
`getStage` is reading an `AtomicReference` and we're doing so four times (including the time within `ensureNotAborted`) so if we do ever log this we will find it hard to be certain that we got the same value each of those times. Could we move things around to avoid doing multiple reads?

Reply:
Done 👍